swift/Sources/Lexer/Lexer.swift

```diff
@@ -9,72 +6,55 @@ extension Character {
 
 extension String {
     func numberString(from index: String.Index) -> String {
-
         var endIndex = self.index(after: index)
-
-        while endIndex < self.endIndex {
-            if self[endIndex].isNumber == false {
-                return String(self[index..<endIndex])
-            }
+        while endIndex < self.endIndex && self[endIndex].isNumber {
             endIndex = self.index(after: endIndex)
         }
-        return String(self[index...])
+        return String(self[index..<endIndex])
     }
 
     func identifier(from index: String.Index) -> String {
         var endIndex = self.index(after: index)
-
-        while endIndex < self.endIndex {
-            if self[endIndex].isValidIdentifier == false {
-                return String(self[index..<endIndex])
-            }
+        while endIndex < self.endIndex && self[endIndex].isValidIdentifier {
             endIndex = self.index(after: endIndex)
         }
-        return String(self[index...])
+        return String(self[index..<endIndex])
     }
 }
 
 struct Lexer {
 
     let input: String
-    var currentIndex: String.Index
-
-    init(input: String) {
-        self.input = input
-        self.currentIndex = input.startIndex
+    lazy var currentIndex = input.startIndex
```

Comment:
Thank you for taking a look at my PR and for your thoughts.  I really liked the changes you had made earlier.  My PR originally had all of the same change recommendations - it was cool to see someone had the same ideas I was having. 😀

The benefit for this line of code is at the point of initialization of a `Lexer`.  By making `currentIndex` a `lazy var`, the initialization of a `Lexer` becomes:

```Swift
var lexer = Lexer(input: <source>)
```

without the additional initializer definition.  While the following line of code you mentioned is a find builder function

``` Swift
static func from(code: String, keyWords: [String: Token] = .defaultKeyWords) -> Lexer {
        Lexer(input: code, keyWords: keyWords, currentIndex: code.startIndex)
    }
```

it also adds unnecessary boiler plate.  I would offer the additional boiler plate does not add benefit or clarification at the point of user.  It would make more sense if were talking about a type that can build a `Lexer`, a `Parser`, an `AST`, or a mix of the three depending upon the builder functions called on the constructing type.

I like the simplicity the one line provides, but if you prefer to have an initializer or builder function, I'll remove this change from the PR.